src/browser_action/on-each-interaction.js

```diff
@@ -11,38 +11,133 @@
+
+    // Go though the list of all events (not just those with interactionIds)
+    // to get details of what was in the longest frame and the earliest and latest event timestamps
+    for (const entry of allEvents) {
+        const renderTime = entry.startTime + entry.duration;
+        // If a previous render time is within 8ms of the largest render time,
+        // assume they were part of the same frame so include them
+        if (Math.abs(renderTime - largestRenderTime) <= 8) {
+          longestFrameEntries.push(entry);
+          groupStartTime = groupStartTime ? Math.min(entry.startTime, groupStartTime) : entry.startTime;
+          groupProcessingStart = groupProcessingStart ?
+            Math.min(
+              entry.processingStart,
+              groupProcessingStart,
+            ) :
+            entry.processingStart;
+          groupProcessingEnd = groupProcessingEnd ? Math.max(entry.processingEnd, groupProcessingEnd) : entry.processingEnd;
+        }
     }
+
+    // For the breakdowns we need to know the first and last interaction events (i.e. those with interactionIds) in that longest list
+    const longestFrameInteractionEntries = longestFrameEntries.filter((entry) => entry.interactionId);
+    // Entries should be in order so can use that to find first
+    const firstInteractionEntry = longestFrameInteractionEntries[0];
+    const firstEventEntry = longestFrameEntries[0];
+    const lastEventEntry = longestFrameEntries.slice(-1);
+
+    // Filter further to get the entries entry (the longest interactionId events with the longest durations)
+    const longestInteractionEntries = longestFrameInteractionEntries.filter((entry) => entry.interactionId  === longestEntry.interactionId && entry.duration === largestDuration);
+
+    // Sometimes target is not set so look across all events with that interactionId to get it
+    const firstInteractionEntryWithTarget = interactions.find((entry) => longestEntry.interactionId === longestEntry.interactionId && entry.target);
+
+    // Filter down LoAFs to ones that intersected any event startTime and any processingEnd
+    // The LoAF processing the last script in that frame will then present that frame so
+    // we get rAF and other rendering work too with this interaction.
+    const longAnimationFrameEntries = getIntersectingLoAFs(groupStartTime, groupProcessingEnd)
+
+    // Since entry durations are rounded to the nearest 8ms, we need to clamp
+    // the `nextPaintTime` value to be higher than the `groupProcessingEnd` or
+    // end time of any LoAF entry.
+    const nextPaintTimeCandidates = [
+      firstEventEntry.startTime + firstEventEntry.duration,
+      groupProcessingEnd,
+    ].concat(
+      longAnimationFrameEntries.map((loaf) => loaf.startTime + loaf.duration),
+    );
+    const nextPaintTime = Math.max.apply(Math, nextPaintTimeCandidates);
+
+    // Gather that all in a similar structure to an INP attribution object and send it back
+    callback({
```

Comment:
The current version of the extension logs on each interaction, but this code will only log once per PerformanceObserver callback (which could contain multiple interactions). Is this intentional?

Reply:
Not really, but at the same time it will mean only the worst interaction in that callback will be observed which is probably not a bad thing.

The main thing is to measure this sort of thing:
1. 50ms interaction
2. 250ms interaction - reporting as INP
3. 249ms interaction
4. 75ms interaction

We don't really care about 1. and 4. to be honest. But not knowing that 3. is a slow interaction despite it only being 1ms slower than the INP one is probably worth knowing.

Having only INP instead of also this interaction is kinda like have the Interactions Track in DevTools and only seeing the largest one (or at least increasing ones). Having interactions logging allows you to see other big ones. So if it occasionally skips shorter ones that's not too bad as you'll probably not even care about those. In fact I'd argue it could be good for reducing the noise of this feature!

For live tracing you kinda want to see the interactions console logs appear as you're interaction. Seeing nothing, and then 3 come in at once isn't really that useful nor how it's intended to be used in most cases. So in many cases with actual distinct interactions this won't be an issue and in others (e.g. keyboard typing) you probably care less about each individual keypress and more about the worst typing experience.

Reply:
> We don't really care about 1. and 4. to be honest. But not knowing that 3. is a slow interaction despite it only being 1ms slower than the INP one is probably worth knowing.

To be clear, the current code, which only reports one interaction per PO callback, does risk not highlighting (3). And it also risks reporting (3) incorrectly (see the other comment thread).

Based on my testing a PO callback will frequently contain both:
- Events from more than one interaction or frame
- Only a subset of events from an interaction or frame

I've been talking to @mmocny about the possibility of Chrome changing this behavior, but currently I do not think it's safe to use PO callbacks as any sort of meaningful boundary or grouping.